File: src/core/request-upload.ts
import type { FormDataFile } from "@literate.ink/utilities";
import type { SessionHandle } from "~/models";
import { aesKeys } from "~/api/private/keys";
import { AES } from "~/api/private/aes";

export class RequestUpload {
  public order: string;
  public id = `selecfile_1_${Date.now()}`;

  private url: string;
  private form: FormData;
  private headers: Record<string, string> = {};

  public constructor (
    private session: SessionHandle,
    public functionName: string,
    file: FormDataFile,
    public fileName: string
  ) {
    session.information.order++;

    const { iv, key } = aesKeys(session);
    this.order = AES.encrypt(session.information.order.toString(), key, iv);

    const form = new FormData();
    form.append("numeroOrdre", this.order);
    form.append("numeroSession", session.information.id.toString());
    form.append("nomRequete", functionName);
    form.append("idFichier", this.id);
    form.append("md5", "");
    // @ts-expect-error : trust me.
    form.append("files[]", file, fileName);

    this.form = form;
    this.url = session.information.url + `/uploadfilesession/${session.information.accountKind}/${session.information.id}`;
    this.headers["Content-Disposition"] = `attachment; filename="${encodeURI(fileName)}"`;
  }

  public async send (): Promise<void> {
    // Please, see the following comment to understand.
    // https://github.com/oven-sh/bun/issues/7917#issuecomment-1872454367
    const textBody = await new Response(this.form).text();
    this.headers["Content-Type"] = `multipart/form-data; boundary=${textBody
      .split("\n")[0]
      .slice(2)}`;

    let state = 3; // Set to UPLOADING by default.

    while (state === 3) { // UPLOADING
      const response = await this.session.fetcher({
        url: new URL(this.url),
        method: "POST",
        content: textBody,
        headers: this.headers
      });

      const json = JSON.parse(response.content);
      state = json.etat;
    }

    // Even if there's an error, it bumped.
    this.session.information.order++;

    if (state === 0) { // UNKNOWN
      throw new Error("The upload state is unknown.");
    }
    else if (state === 2) { // ERROR
      throw new Error("The upload failed.");
    }
  }
}

File: library/src/commonMain/kotlin/ink/literate/pawnote/core/request-upload.kt
package ink.literate.pawnote.core

import ink.literate.pawnote.api.private.AES
import ink.literate.pawnote.api.private.aesKeys
import ink.literate.pawnote.models.SessionHandle
import ink.literate.pawnote.models.errors.UploadFailedError
import io.ktor.client.request.*
import io.ktor.client.request.forms.*
import io.ktor.http.*
import io.ktor.http.content.*
import io.ktor.util.*
import io.ktor.utils.io.*
import java.io.File
import java.nio.file.Files
import kotlinx.datetime.Clock
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.int
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

@OptIn(InternalAPI::class)
class RequestUpload(
    private val session: SessionHandle,
    functionName: String,
    file: File,
    private val fileName: String
) {
  val id = "selecfile_1_${Clock.System.now().toEpochMilliseconds()}"
  var order: String = ""

  private var url = ""
  private var form: List<PartData>? = null

  init {
    session.information.order++

    val keys = aesKeys(session.information)

    order = AES.encrypt(session.information.order.toString().toByteArray(), keys.key, keys.iv)

    val form = formData {
      append("\"numeroOrdre\"", order)
      append("\"numeroSession\"", session.information.id.toString())
      append("\"nomRequete\"", functionName)
      append("\"idFichier\"", id)
      append("\"md5\"", "")
      append(
          "\"files[]\"",
          file.readBytes(),
          Headers.build {
            append(HttpHeaders.ContentDisposition, "filename=\"${fileName}\"")
            append(HttpHeaders.ContentType, Files.probeContentType(file.toPath()))
          })
    }

    this.form = form
    this.url =
        session.information.url +
            "/uploadfilesession/${session.information.accountKind.code}/${session.information.id}"
  }

  suspend fun send() {
    var state = 3

    val url = this.url

    while (state == 3) {
      val response =
          this.session.client.submitFormWithBinaryData(url = url, form!!) {
            headers {
              append(HttpHeaders.ContentDisposition, "attachment; filename=\"${fileName}\"")
            }
          }

      val json = Json.parseToJsonElement(response.content.readUTF8Line()!!)
      state = json.jsonObject["etat"]!!.jsonPrimitive.int
    }

    // Even if there's an error, it bumped.
    this.session.information.order++

    if (state == 0 || state == 2) throw UploadFailedError()
  }
}
